### ISSUE TITLE
publishing multiple messages within the context of a received message

### ISSUE THREAD

**Issue by shot-codes** [libp2p/rust-libp2p#2649]
## Summary
I am publishing multiple times within the match of a received message. All messages do get sent but only after the VecDeque is emptied. My understanding is that this is because the swarm is not progressed by the select! macro until after I leave the match arm. Is there any way to explicitly tell the swarm to progress after each publish such that they are sent as they are iterated through, rather than all at once afterwards?
```rust
loop {
    select! {
        event = swarm.select_next_some() => match event {
            SwarmEvent::Behaviour(GossipsubEvent::Message {
                propagation_source: _peer_id,
                message_id: _id,
                message,
            }) => {
                match message.topic.as_str() {

                    "some_topic" => {

                        let mut pois = VecDeque::from(["a", "b", "c", "d"]);

                        while !pois.is_empty() {
                            let data = pois.pop_front();
                            if let Err(e) = swarm
                                .behaviour_mut()
                                .publish(some_other_topic.clone(), data.as_bytes())
                            {
                                println!("Publish error: {:?}", e);
                            };
                            std::thread::sleep(Duration::from_millis(1000));
                        }

                    }
                }
            }
        }
    }
}
```
## Expected behaviour

Messages are published as soon as I call `swarm.behaviour_mut().publish()`

## Actual behaviour

Messages are first published once the match arm is exited.

## Version

- libp2p version (version number, commit, or branch): 0.44.0

**Comment by mxinden** [comment]
How about a `continue` statement for the outer loop?

**Comment by thomaseizinger** [comment]
> Messages are published as soon as I call `swarm.behaviour_mut().publish()`

The issue is that you need to call `Swarm::select_next_some` again before we can process the state changes within the gossipsub behaviour.

If you want to publish messages at an interval, I'd suggest to spawn a separate task and read from a channel connected to that task and call publish for each message received from it. See https://github.com/libp2p/rust-libp2p/blob/master/examples/file-sharing/src/network.rs for something similar.
